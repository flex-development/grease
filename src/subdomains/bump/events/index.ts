/**
 * @file Entry Point - Events
 * @module grease/bump/events
 */

export { default as BumpEvent } from './bump.event'
export { default as BumpEventListener } from './bump.listener'
