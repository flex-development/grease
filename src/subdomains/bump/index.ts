/**
 * @file Entry Point - bump
 * @module grease/bump
 */

export { default as BumpModule } from './bump.module'
export * from './events'
export * from './models'
export * from './operations'
export * from './queries'
