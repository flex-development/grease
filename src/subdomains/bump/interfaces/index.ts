/**
 * @file Entry Point - Interfaces
 * @module grease/bump/interfaces
 */

export type { default as IRecommendedBump } from './recommended-bump.interface'
