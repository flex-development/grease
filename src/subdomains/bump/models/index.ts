/**
 * @file Entry Point - Models
 * @module grease/bump/models
 */

export { default as RecommendedBump } from './recommended-bump.model'
