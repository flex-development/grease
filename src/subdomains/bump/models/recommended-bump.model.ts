/**
 * @file Models - RecommendedBump
 * @module grease/bump/models/RecommendedBump
 */

import { ReleaseType } from '#src/enums'
import { ifelse, type Omit } from '@flex-development/tutils'

/**
 * Recommended version bump.
 *
 * @class
 */
class RecommendedBump {
  /**
   * Total number of breaking changes committed since last release.
   *
   * @public
   * @instance
   * @member {number} breaks
   */
  public breaks: number

  /**
   * Recommended release type.
   *
   * @see {@linkcode ReleaseType}
   *
   * @public
   * @instance
   * @member {Extract<ReleaseType, 'major' | 'minor' | 'patch'>} bump
   */
  public bump: Extract<ReleaseType, 'major' | 'minor' | 'patch'>

  /**
   * Number of commits considered.
   *
   * @public
   * @instance
   * @member {number} commits
   */
  public commits: number

  /**
   * Total number of features committed since last release.
   *
   * @public
   * @instance
   * @member {number} features
   */
  public features: number

  /**
   * Create a new version bump recommendation.
   *
   * @param {Omit<RecommendedBump, 'bump'>} data - Version bump data
   */
  constructor(data: Omit<RecommendedBump, 'bump'>) {
    this.breaks = data.breaks
    this.commits = data.commits
    this.features = data.features
    this.bump = ifelse(
      this.breaks,
      ReleaseType.MAJOR,
      ifelse(this.features, ReleaseType.MINOR, ReleaseType.PATCH)
    )
  }
}

export default RecommendedBump
