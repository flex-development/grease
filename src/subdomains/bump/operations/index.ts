/**
 * @file Entry Point - Operations
 * @module grease/bump/operations
 */

export { default as BumpOperationHandler } from './bump.handler'
export {
  default as BumpOperation,
  type BumpOperationDTO
} from './bump.operation'
