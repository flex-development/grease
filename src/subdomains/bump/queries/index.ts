/**
 * @file Entry Point - Queries
 * @module grease/bump/queries
 */

export { default as BumpQueryHandler } from './bump.handler'
export { default as BumpQuery } from './bump.query'
