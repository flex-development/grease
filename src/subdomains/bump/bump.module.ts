/**
 * @file BumpModule
 * @module grease/bump/BumpModule
 */

import { LoggerService } from '#src/providers'
import { Module } from '@nestjs/common'
import { BumpEventListener } from './events'
import { BumpOperationHandler } from './operations'
import { BumpQueryHandler } from './queries'

/**
 * Bump operations module.
 *
 * @class
 */
@Module({
  providers: [
    BumpEventListener,
    BumpOperationHandler,
    BumpQueryHandler,
    LoggerService
  ]
})
class BumpModule {}

export default BumpModule
