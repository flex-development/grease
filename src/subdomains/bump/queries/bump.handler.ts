/**
 * @file Queries - BumpQueryHandler
 * @module grease/bump/queries/BumpQueryHandler
 */

import { RecommendedBump } from '#src/bump/models'
import { CommitQuery, TagQuery, type ICommit } from '#src/git'
import { ValidationService } from '#src/providers'
import { at } from '@flex-development/tutils'
import { QueryBus, QueryHandler, type IQueryHandler } from '@nestjs/cqrs'
import BumpQuery from './bump.query'

/**
 * Recommended version bump query handler.
 *
 * @see {@linkcode BumpQuery}
 * @see {@linkcode RecommendedBump}
 *
 * @class
 * @implements {IQueryHandler<BumpQuery,RecommendedBump>}
 */
@QueryHandler(BumpQuery)
class BumpQueryHandler implements IQueryHandler<BumpQuery, RecommendedBump> {
  /**
   * Create a new recommended version bump query handler.
   *
   * @param {QueryBus} queries - Query bus
   * @param {ValidationService} validator - Validation service
   */
  constructor(
    protected readonly queries: QueryBus,
    protected readonly validator: ValidationService
  ) {}

  /**
   * Execute a version bump recommendation query.
   *
   * @see {@linkcode BumpQuery}
   * @see {@linkcode RecommendedBump}
   *
   * @public
   * @async
   *
   * @param {BumpQuery} query - Query to execute
   * @return {RecommendedBump} Recommended version bump
   */
  public async execute(query: BumpQuery): Promise<RecommendedBump> {
    await this.validator.validate(query)

    /**
     * Commits since last release.
     *
     * @const {ICommit[]} commits
     */
    const commits: ICommit[] = await this.queries.execute(new CommitQuery({
      ...query,
      from: at(await this.queries.execute(new TagQuery(query)), 0, '')
    }))

    /**
     * Total number of breaking changes committed since last release.
     *
     * @const {number} breaks
     */
    const breaks: number = commits.reduce((acc, commit) => {
      return acc + (commit.breaking ? 1 : 0)
    }, 0)

    /**
     * Total number of features committed since last release.
     *
     * @const {number} features
     */
    const features: number = commits.reduce((acc, commit) => {
      return acc + (/^feat(ure)?/.test(commit.type) ? 1 : 0)
    }, 0)

    return new RecommendedBump({ breaks, commits: commits.length, features })
  }
}

export default BumpQueryHandler
