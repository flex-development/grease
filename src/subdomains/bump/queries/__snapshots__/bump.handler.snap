// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:bump/queries/BumpQueryHandler > #execute > should return recommended version bump 1`] = `
RecommendedBump {
  "breaks": 1,
  "bump": "major",
  "commits": 26,
  "features": 7,
}
`;
